### PR TITLE
Make BasicInformationCluster use AttributeAccessInterface to save RAM

### PR DESCRIPTION
**Pull request by tcarmelveilleux**
#### Problem

BasicInformationCluster round-trips through RAM with every attribute currently "RAM" configured. Since most values of BasicInformationCluster are constants that are fixed for an entire build, and that they are currently all returned by ConfigurationManager accessor, the round trips to RAM take more space than required and prevent using the direct source of truth for when values change.

Issue #9261

#### Change overview

- Make BasicInformation Cluster employ AttributeAccessInterface
- Move all fixed attributes to External storage
- Better document the AttributeAccessInterface fallback path

#### Testing
- Unit tests still pass
- Integration tests still pass

### CODE REVIEW

File: src/app/clusters/basic/basic.cpp [hunk 1]
```diff
@@ -62,41 +62,251 @@ CHIP_ERROR BasicAttrAccess::Read(const ConcreteReadAttributePath & aPath, Attrib
...
-    else
+
+    if (status == CHIP_ERROR_NOT_FOUND)
     {
-        err = aEncoder.Encode(chip::CharSpan("XX", strlen("XX")));
+        // Somehow attribute support not handled by custom logic. Until all such
+        // paths are resolved, allow fallback to attribute store or other heuristics
+        // which may be done by callers of this method.
+        status = CHIP_NO_ERROR;
     }
```
Comment (msandstedt) [inline src/app/clusters/basic/basic.cpp:292]: I thought we had a concept of handled / not handled, which then I thought could fallback to some other handler (attribute store?) for not-handled.  Can something like that be leveraged here to clean this up?  Or is that in fact what we are already doing?  If so, why do we need to translate here from CHIP_ERROR_NOT_FOUND to CHIP_NO_ERROR?  Why can't the path where something is not found just return CHIP_NO_ERROR if we intend that this is not an error?

File: src/app/clusters/basic/basic.cpp [hunk 2]
```diff
@@ -62,41 +62,251 @@ CHIP_ERROR BasicAttrAccess::Read(const ConcreteReadAttributePath & aPath, Attrib
...
+        break;
+
+    case VendorName::Id: {
+        constexpr size_t kMaxLen     = DeviceLayer::ConfigurationManager::kMaxVendorNameLength;
+        char vendorName[kMaxLen + 1] = { 0 };
+        status                       = ConfigurationMgr().GetVendorName(vendorName, sizeof(vendorName));
+        if (status == CHIP_NO_ERROR)
+        {
+            status = aEncoder.Encode(chip::CharSpan(vendorName, strnlen(vendorName, kMaxLen)));
+        }
```
Comment (msandstedt) [inline src/app/clusters/basic/basic.cpp:80]: It seems to be the preference of the project to use the goto-wrapper and early return macros to reduce the number of dedicated error-checking if statements.

I don't know if any of those can be leveraged here.  Or perhaps a new macro could be introduced.  But there is a lot of this `if (status == CHIP_NO_ERROR) { // do next thing }` construct in this file.

File: examples/lighting-app/lighting-common/lighting-app.matter [hunk 3]
```diff
@@ -957,6 +957,21 @@ client cluster OnOff = 6 {
   attribute enum8 startUpOnOff = 16387;
   readonly global attribute bitmap32 featureMap = 65532;
   readonly global attribute int16u clusterRevision = 65533;
+
+  request struct OffWithEffectRequest {
```
Comment (bzbarsky-apple) [inline examples/lighting-app/lighting-common/lighting-app.matter:961]: These look like spurious changes.

File: src/app/clusters/basic/basic.cpp [hunk 4]
```diff
@@ -62,41 +62,251 @@ CHIP_ERROR BasicAttrAccess::Read(const ConcreteReadAttributePath & aPath, Attrib
...
+
+        if (status == CHIP_NO_ERROR)
+        {
+            // Format is YYYYMMDD
+            snprintf(manufacturingDateString, sizeof(manufacturingDateString), "%04" PRIu16 "%02" PRIu8 "%02" PRIu8,
+                     manufacturingYear, manufacturingMonth, manufacturingDayOfMonth);
+            status = aEncoder.Encode(chip::CharSpan(manufacturingDateString, strnlen(manufacturingDateString, kMaxLen)));
+        }
+        ChipLogError(Zcl, "MAAAAANNNNUFFF: %s, err: %" CHIP_ERROR_FORMAT, manufacturingDateString, status.Format());
```
Comment (bzbarsky-apple) [inline src/app/clusters/basic/basic.cpp:181]: I don't think we want the unconditional ChipLogError here.

File: src/app/clusters/basic/basic.cpp [hunk 5]
```diff
@@ -62,41 +62,251 @@ CHIP_ERROR BasicAttrAccess::Read(const ConcreteReadAttributePath & aPath, Attrib
...
         {
-            err = aEncoder.Encode(chip::CharSpan(location, strlen(location)));
+            status = aEncoder.Encode(chip::CharSpan(uniqueId, strnlen(uniqueId, kMaxLen)));
         }
+        break;
+    }
+
+    default:
+        // We did not find a processing path, the caller will delegate elsewhere.
+        status = CHIP_ERROR_NOT_FOUND;
```
Comment (bzbarsky-apple) [inline src/app/clusters/basic/basic.cpp:282]: Why not just set status to `CHIP_NO_ERROR`?

File: src/app/clusters/basic/basic.cpp [hunk 6]
```diff
@@ -62,41 +62,251 @@ CHIP_ERROR BasicAttrAccess::Read(const ConcreteReadAttributePath & aPath, Attrib
...
+    return status;
+}
+
+CHIP_ERROR BasicAttrAccess::ReadLocation(AttributeValueEncoder & aEncoder)
+{
+    constexpr size_t kMaxLen   = DeviceLayer::ConfigurationManager::kMaxLocationLength;
+    char location[kMaxLen + 1] = { 0 };
+    size_t codeLen             = 0;
+
+    CHIP_ERROR err = ConfigurationMgr().GetCountryCode(location, sizeof(location), codeLen);
```
Comment (bzbarsky-apple) [inline src/app/clusters/basic/basic.cpp:303]: Does GetCountryCode guarantee null-termination of the output?  Why is it also returning codeLen if it does?

I guess the existing code in the init method was making this assumption too....

File: src/app/util/ember-compatibility-functions.cpp [hunk 7]
```diff
@@ -425,10 +425,29 @@ CHIP_ERROR ReadSingleClusterData(const SubjectDescriptor & aSubjectDescriptor, b
             (attributeCluster != nullptr) ? &reader : findAttributeAccessOverride(aPath.mEndpointId, aPath.mClusterId);
         if (attributeOverride)
         {
-            bool triedEncode;
-            ReturnErrorOnFailure(ReadViaAccessInterface(aSubjectDescriptor.fabricIndex, aIsFabricFiltered, aPath, aAttributeReports,
-                                                        apEncoderState, attributeOverride, &triedEncode));
-            ReturnErrorCodeIf(triedEncode, CHIP_NO_ERROR);
+            bool triedEncode = false;
```
Comment (bzbarsky-apple) [inline src/app/util/ember-compatibility-functions.cpp:428]: Why the changes here?  The logic stays the same, right?  Is it just to document the API contract for AttributeAccessInterface::Read which is already documented on that interface?

File: src/app/clusters/basic/basic.cpp (hunk 4 above)
Comment (tcarmelveilleux) [reply src/app/clusters/basic/basic.cpp:181]: Sorry, was debuggign sometjing from CI failure and this got left behind.

File: src/app/util/ember-compatibility-functions.cpp (hunk 7 above)
Comment (tcarmelveilleux) [reply src/app/util/ember-compatibility-functions.cpp:428]: One of my in-between changes tried to handle the fact that something is found versus not found to be supported. This goes back to optionality where an optional attribute may only be known to be optional from runtime support of some underlying data accessor. I can revert all of this now I guess, but it was *really not obvious* without a lot of reverse engineering, how the code paths on success/error yielded different outcomes.

File: src/app/clusters/basic/basic.cpp (hunk 6 above)
Comment (tcarmelveilleux) [reply src/app/clusters/basic/basic.cpp:272]: I did not check into that yet. The API is overall really poor :( It's mixed between null and non-null terminated strings in ConfigurationManager

File: src/app/clusters/basic/basic.cpp (hunk 5 above)
Comment (tcarmelveilleux) [reply src/app/clusters/basic/basic.cpp]: Fixed

File: examples/lighting-app/lighting-common/lighting-app.matter (hunk 3 above)
Comment (tcarmelveilleux) [reply examples/lighting-app/lighting-common/lighting-app.matter]: Was due to codegen non-determinism

File: src/app/clusters/basic/basic.cpp [hunk 8]
```diff
@@ -54,6 +54,12 @@ class BasicAttrAccess : public AttributeAccessInterface
 
 BasicAttrAccess gAttrAccess;
 
+CHIP_ERROR EncodeStringOnSuccess(CHIP_ERROR status, AttributeValueEncoder & encoder, const char * buf, size_t maxBufSize)
```
Comment (carol-apple) [inline src/app/clusters/basic/basic.cpp:57]: nit: Make this a static function

File: src/app/clusters/basic/basic.cpp [hunk 9]
```diff
@@ -62,41 +68,214 @@ CHIP_ERROR BasicAttrAccess::Read(const ConcreteReadAttributePath & aPath, Attrib
...
+            manufacturingDayOfMonth = 1;
+            status                  = CHIP_NO_ERROR;
         }
+
+        if (status == CHIP_NO_ERROR)
+        {
+            // Format is YYYYMMDD
+            snprintf(manufacturingDateString, sizeof(manufacturingDateString), "%04" PRIu16 "%02" PRIu8 "%02" PRIu8,
+                     manufacturingYear, manufacturingMonth, manufacturingDayOfMonth);
+            status = aEncoder.Encode(chip::CharSpan(manufacturingDateString, strnlen(manufacturingDateString, kMaxLen)));
```
Comment (carol-apple) [inline src/app/clusters/basic/basic.cpp:173]: nit: Just call `EncodeStringOnSuccess` directly?

File: src/app/clusters/basic/basic.cpp (hunk 8 above)
Comment (tcarmelveilleux) [reply src/app/clusters/basic/basic.cpp:57]: It is in the anonymous namespace, it is thus static

File: src/app/util/ember-compatibility-functions.cpp (hunk 7 above)
Comment (tcarmelveilleux) [reply src/app/util/ember-compatibility-functions.cpp]: Removed

File: src/app/clusters/basic/basic.cpp (hunk 5 above)
Comment (tcarmelveilleux) [reply src/app/clusters/basic/basic.cpp]: Fixed

**Review by bzbarsky-apple — APPROVED** [review]

File: src/app/clusters/basic/basic.cpp (hunk 9 above)
Comment (tcarmelveilleux) [reply src/app/clusters/basic/basic.cpp:176]: Done

**Review by carol-apple — APPROVED** [review]

**Review by woody-apple — APPROVED** [review]

**Review by cecille — APPROVED** [review]

**Review by jmartinez-silabs — APPROVED** [review]
